pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/TTLPartitionUpsertMetadataManager.java

```diff
@@ -0,0 +1,68 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.segment.local.upsert;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.ThreadSafe;
+import org.apache.pinot.common.metrics.ServerMetrics;
+import org.apache.pinot.segment.local.utils.HashUtils;
+import org.apache.pinot.segment.spi.IndexSegment;
+import org.apache.pinot.spi.config.table.HashFunction;
+
+
+/**
+ * Implementation of {@link PartitionUpsertMetadataManager} that is backed by a {@link ConcurrentHashMap}.
+ * Used by Upsert TTL (Time-to-live) feature.
+ */
+@SuppressWarnings({"rawtypes", "unchecked"})
+@ThreadSafe
+public class TTLPartitionUpsertMetadataManager extends ConcurrentMapPartitionUpsertMetadataManager {
```

Comment:
why is this class needed when you can just override the method `removeExpiredPrimaryKeys` in `ConcurrentMapPartitionUpsertMetadataManager`? 
Same question on `TTLTableUpsertMetadataManager`.

Reply:
@navina I will add more methods later in the TTLPartitionUpsertMetadataManager.

As mentioned in https://github.com/apache/pinot/issues/9529,

- When committing segment, update replaceSegment to clean up keys;
- when loading segments, get snapshot from peer-server to avoid re-compute

Reply:
You can add more methods to it. But my point was that this TTL feature will only work with `ConcurrentMapPartitionUpsertMetadataManager` and none of the other implementations of `PartitionUpsertMetadataManager`. I don't think that was the intention in the design. Please correct me if I am missing something here. 

It will be better to model this as a separate component or a wrapper around any of the `PartitionUpsertMetadataManager`.

Reply:
+1 here. I think TTL should be a feature supported by general upsert metadata manager. Making it a separate class will prevent other storage format from supporting TTL

Reply:
thanks @navina and @Jackie-Jiang Let me refactor this part.